### PR TITLE
[CR] Ensure player::on_dodge is actually called

### PR DESCRIPTION
**Pull request by mugling**
It appears `player::on_dodge` is never called.

`monster::on_dodge` is called from `monster::deal_melee_attack`. As we don't override `deal_melee_attack` for `player` the base method `creature::deal_melee_attack` is used which does not call `on_dodge`.

This PR changes this so that the appropriate `on_dodge` method is called by `player::melee_attack` or `monster::melee_attack` and removes `monster::deal_melee_attack` which becomes unnecessary after this change.

Training of the dodging skill is moved from `on_hit` to `on_dodge` so the player now learns from successful dodges as opposed to being beat on the head.

I'd like to add further logic to `player::on_dodge` but this will be part of a separate PR (#14628) to limit scope.

### CODE REVIEW

**Comment by illi-kun** [comment]
I really like things you're doing!

**Comment by SeanMirrsen** [comment]
Minor thing, but maybe leave some trickle of dodging experience from getting hit? Learn from mistakes, incentive to improve, etc.

File: src/player.cpp [hunk 1]
```diff
@@ -4690,21 +4690,13 @@ void player::on_dodge( Creature *source, int difficulty )
...
     if( source == nullptr || proj != nullptr ) {
         return;
     }
 
-    if( difficulty == INT_MIN ) {
-        difficulty = source->get_melee();
-    }
-
-    if( difficulty > 0 ) {
-        practice( skill_dodge, difficulty );
```
Comment (Coolthulhu) [inline src/player.cpp]: This seems to change things a fair bit:
Before, players hit by attacks would still train dodge, meaning that they could eventually start dodging them.
Is on-hit dodge practice handled somewhere else or is it gone?

**Comment by chaosvolt** [comment]
I would be in favor of failed dodges still adding experience, yes.

Also, this leads me to a question. Do martial arts buff that proc on dodging go off right? I'm unsure where that is checked for.

**Comment by mugling** [comment]
I suppose training on failed dodges is required otherwise how would you practice the skill in the first place so as to be able to dodge successfully?

Do I revert ce3386b (so training occurs on both dodges and hits) or do we implement something different - perhaps granting different experience for dodges vs hits?

> Do martial arts buff that proc on dodging go off right

Previously I don't think so as `player::ma_ondodge_effects` is called from `player::on_dodge` but the changes in e722219 should fix that.

**Comment by chaosvolt** [comment]
Hmm. I might suggest one of them give...I guess half the experience. But I'm unsure which should be halved, failed dodges or successful ones.

**Comment by Coolthulhu** [comment]
> perhaps granting different experience for dodges vs hits?

How about halving the "perceived" difficulty in the `on_hit` section? That way armored, encumbered tin cans would hit their cap pretty fast, while unencumbered low-skill survivors would only get slower progression early on.

This would make grinding dodge by getting armored and surrounded inefficient, except for the very early levels which you should already have by the time you can do the whole armored grind routine.

**Comment by SeanMirrsen** [comment]
I'm not sure how the dodge check works, but I assume it involves a random roll. What if the experience gain is scaled proportionally to how *close* the check was to succeeding or failing? Say, the dodge value to beat is 10. Rolling a 9 barely fails, rolling a 10 barely succeeds. So a roll of either 9 or 10 would give full experience. An 8 or 11 would give 90%, a 7 or 12 would give 80%, etc. I don't know the scale of numbers involved, but there should be a curve - the further you are from the target number, you gain increasingly less experience. So that either you're too bad to learn dodging since you're not dodging at all, or you're too good and you're not learning anything.

**Comment by Coolthulhu** [comment]
`on_dodge` doesn't have access to the roll, only to its difficulty.
Tying it to roll precision like that would also make it so that wearing heavy armor would allow training dodge on weak stuff with no repercussions.

**Comment by SeanMirrsen** [comment]
>Tying it to roll precision like that would also make it so that wearing heavy armor would allow training dodge on weak stuff with no repercussions.

That makes sense. But I'm guessing that's a bad thing ingame. Maybe also scale it proportionally to how difficult the attack being dodged is to dodge as well? Like, set the maximum xp you can gain by the difficulty, set the percentage of the maximum gained by the margin of success or failure.

**Comment by Coolthulhu** [comment]
> Like, set the maximum xp you can gain by the difficulty, set the percentage of the maximum gained by the margin of success or failure.

Just having scaling according to difficulty is enough, no need to have to weight self down with armor.
Especially considering how wearing armor is already a very important thing - even if only successful dodges granted xp, dodge training would still involve a complete set of armor in most cases.

**Comment by SeanMirrsen** [comment]
Scaling by difficulty without scaling for margin of success, with failure also giving xp, would result in loading up on armor and standing there tanking hits from something small and fast to "train". It needs to give less xp as you fail more.

**Comment by Coolthulhu** [comment]
> It needs to give less xp as you fail more.

But it shouldn't give any more xp for voluntarily slowing self down either.

Halving the difficulty on failed dodge would provide a cap on skill gains from tanking without needlessly penalizing unarmored or weak survivors.
Highest skill for a grindable enemy is 10 (manhack). Halved difficulty on failed dodge would only allow grinding manhacks up to 5 skill level.
It wouldn't prevent early lab survivors from gaining skill from getting shredded by those manhacks, though.

**Comment by SeanMirrsen** [comment]
>But it shouldn't give any more xp for voluntarily slowing self down either.

In given, specific circumstances - why not? People wear weighted clothing for all kinds of training to add to the resistance they have to overcome and get even better results.

**Comment by Coolthulhu** [comment]
> In given, specific circumstances - why not? People wear weighted clothing for all kinds of training to add to the resistance they have to overcome and get even better results.

In case of training dodging, it would be much more likely to disrupt the training than to help it.
Changes in balance due to uneven weight distribution, ease of blocking rather than dodging, extra effort required to perform the dodge combined with decreased penalty for not dodging etc.

And it would also disrupt the other kind of balance: the one in game. There is no reason to penalize ultra-lightly armored survivors compared to moderately-lightly armored ones.
If anything, there should be more incentive to drop the armor. Or more incentive to go fully armored. Anything but the current limited optimum, which is going light enough to dodge hulks but heavy enough to ignore anything less than spitters.

**Comment by mugling** [comment]
@Coolthulhu I think your suggestion is the most appropriate. I've done some further refactoring in #14636 (which in effect gives equal training on both hits and misses). Once that PR is merged I'll submit a further for the behavioral changes to reduce the perceived difficulty of failed dodges.

**Comment by kevingranade** [comment]
Regarding practice rates in general, the way I want to take it in the
future is having a flat amount of practice per time spent exercising the
skill, with the only discrimination being what tasks are considered
significant for practicing at the current level.
The idea of having different actions advance you at different rates is what
leads to bizarre activities being optimal for levelling certain skills.

**Comment by SeanMirrsen** [comment]
Well, if that's the way you want it. :P 
In my opinion, "bizarre activities" is par for the course as far as targeted skill training goes.
http://www.taodewan.com/WCkungfu/images/sifu-jim-fung-wooden-dummy-l.jpg
:P

**Comment by kevingranade** [comment]
If you've studied that style of kung-fu at all, those dolls make perfect
sense :P

**Comment by chaosvolt** [comment]
Making both success and failure train dodge equally probably would be best to discourage oddball training tactics, yeah. At least in this particular instance.

**Pull request closed without merging**